apt/risk/data_assessment/attack_strategy_utils.py

```diff
@@ -74,3 +87,91 @@ def find_knn(self, knn_learner: NearestNeighbors, query_samples: ArrayDataset, d
+            except ValueError as e:
+                if str(e).startswith('For each axis slice, the sum of'):
+                    print('Column', column, e)
+                else:
+                    raise
+        elif is_numeric:
+            if test_type == 'KS':
+                result = stats.ks_2samp(df1_column_samples, df2_column_samples)
+            elif test_type == 'CVM':
+                result = stats.cramervonmises_2samp(df1_column_samples, df1_column_samples)
+            else:
+                raise ValueError('Unknown test type', test_type)
+        else:
+            print(f'Skipping non-numeric and non-categorical column {column}')
+            return
+        print(
+            f"{column}: {test_type} = {result.statistic:.4f} "
+            f"(p-value = {result.pvalue:.3e}, are equal = {result.pvalue > 0.05})")
+        if result.pvalue < alpha:
+            # Reject H0, different distributions
+            print(f"Distributions differ in column {column}, p-value: {result.pvalue}")
+            differing_columns.append(column)
+        else:
+            # Accept H0, similar distributions
+            print(f'Accept H0, similar distributions in column {column}')
+
+    @staticmethod
+    def _columns_different_distributions(df1: ArrayDataset, df2: ArrayDataset,
+                                         categorical_features: list = [],
+                                         alpha=0.05, test_type='KS') -> list:
+        differing_columns = []
+        df1_samples = df1.get_samples()
+        df2_samples = df2.get_samples()
+        if df1.is_pandas:
+            def is_categorical(col_name):
+                col_name in categorical_features or is_categorical_dtype(df1_samples.dtypes[col_name])
+
+            def is_numeric(col_name):
+                is_numeric_dtype(df1_samples.dtypes[col_name])
+
+            for name, _ in df1_samples.items():
+                KNNAttackStrategyUtils._column_statistical_test(df1_samples[name], df2_samples[name], name,
+                                                                is_categorical, is_numeric(df1_samples.dtypes[name]),
+                                                                test_type, alpha, differing_columns)
+        else:
+            is_df1_numeric_dtype = np.issubdtype(df1_samples.dtype, int) or np.issubdtype(df1_samples.dtype, float)
+
+            def is_categorical(col_name):
+                col_name in categorical_features
+
+            for i, column in enumerate(df1_samples.T):
+                KNNAttackStrategyUtils._column_statistical_test(df1_samples[:, i], df2_samples[:, i], i,
+                                                                is_categorical, is_df1_numeric_dtype, test_type, alpha,
+                                                                differing_columns)
+        return differing_columns
+
+    def validate_distributions(self, original_data_members: ArrayDataset, original_data_non_members: ArrayDataset,
```

Comment:
I thought we said we would externalize alpha to the user as well (but here it's only using the default value).